revanced-patcher/src/main/kotlin/app/revanced/patcher/apk/Apk.kt

```diff
@@ -0,0 +1,277 @@
+import app.revanced.arsc.archive.Archive
+import app.revanced.arsc.resource.ResourceContainer
+import app.revanced.patcher.Patcher
+import app.revanced.patcher.PatcherOptions
+import app.revanced.patcher.logging.asArscLogger
+import app.revanced.patcher.util.ProxyBackedClassList
+import com.reandroid.apk.ApkModule
+import com.reandroid.apk.xmlencoder.EncodeException
+import com.reandroid.arsc.chunk.xml.AndroidManifestBlock
+import com.reandroid.arsc.value.ResConfig
+import lanchon.multidexlib2.*
+import org.jf.dexlib2.Opcodes
+import org.jf.dexlib2.dexbacked.DexBackedDexFile
+import org.jf.dexlib2.iface.DexFile
+import org.jf.dexlib2.iface.MultiDexContainer
+import org.jf.dexlib2.writer.io.MemoryDataStore
+import java.io.File
+
+/**
+ * An [Apk] file.
+ */
+sealed class Apk private constructor(module: ApkModule) {
+    /**
+     * A wrapper around the zip archive of this [Apk].
+     *
+     * @see Archive
+     */
+    private val archive = Archive(module)
+
+    /**
+     * The metadata of the [Apk].
+     */
+    val packageMetadata = PackageMetadata(module.androidManifestBlock)
+
+    val resources = ResourceContainer(archive, module.tableBlock)
+
+    /**
+     * Refresh updated resources and close any open files.
+     *
+     * @param options The [PatcherOptions] of the [Patcher].
+     */
+    internal open fun cleanup(options: PatcherOptions) {
+        try {
+            archive.cleanup(options.logger.asArscLogger())
+        } catch (e: EncodeException) {
+            throw ApkResourceException.Encode(e.message!!, e)
+        }
+
+        resources.refreshPackageName()
+    }
+
+    /**
+     * Write the [Apk] to a file.
+     *
+     * @param output The target file.
+     */
+    fun write(output: File) = archive.save(output)
+
+    companion object {
+        const val MANIFEST_FILE_NAME = "AndroidManifest.xml"
+
+        /**
+         * Determine the [Module] and [Type] of an [ApkModule].
+         *
+         * @return A [Pair] containing the [Module] and [Type] of the [ApkModule].
+         */
+        fun ApkModule.identify(): Pair<Module, Type> {
+            val manifestElement = androidManifestBlock.manifestElement
+            return when {
+                isBaseModule -> Module.Main to Type.Base
+                // The module is a base apk for a dynamic feature module if the "isFeatureModule" attribute is set to true.
+                manifestElement.searchAttributeByName("isFeatureModule")?.valueAsBoolean == true -> Module.DynamicFeature(
+                    split
+                ) to Type.Base
+
+                else -> {
+                    val module = manifestElement.searchAttributeByName("configForSplit")
+                    ?.let { Module.DynamicFeature(it.valueAsString) } ?: Module.Main
+
+                    // Examples:
+                    // config.xhdpi
+                    // df_my_feature.config.en
+                    val config = this.split.split(".").last()
```

Comment:
Are those conventions written in stone or can a language APKModule not end with a two-letter country code?

Reply:
Android uses [ISO 639-1:2002](https://www.iso.org/standard/22109.html) language codes and those are always two letters long.

Language variants (en_US, en_GB) are just stored in one apk (en).